language-plutus-core/src/Language/PlutusCore/TypeRenamer.hs

```diff
@@ -26,6 +76,8 @@ insertName u s = modify (first (IM.insert u s))
 defMax :: Int -> IdentifierM (Maybe BSL.ByteString, Int)
 defMax u = (,) <$> gets (IM.lookup u . fst) <*> gets (fst . IM.findMax . fst)
 
+-- POSSIBLY consider instead of rewriteWith, just adding to an IntMap and then
```

Comment:
Yes, this needs to take scope into account, but we get that essentially for free, because inner bindings shadow outer bindings. Let's suppose we maintain an `IntMap Int`, which we call `renaming`, and let's rename this term (where the numeric suffixes are the uniques and I use "normal" lambda terms, not PLC, to keep it simple)

```
\x_1 -> (\y_2 -> \x_1 -> x_1) x_1
``` 

This is what we get from parsing, where both bindings of `x` carry the same unique. Hence, the inner binding shadows the outer.

We start with `renaming` being empty. Now when we get to the first (outermost) `\x_1 -> `, we check `renaming` and there is no entry for `1`. This indicates that this is the first occurrences of `x`. Hence, we add `1 |-> 1` to `renaming` (i.e., `x_1` remains `x_1`) and proceed to rename the body of this lambda abstraction (i.e., `(\y_2 -> \x_1 -> x_1) x_1`) with `renaming = {1 |-> 1}`. The next node is an application, so we proceed with the same value of `renaming` down each subexpression.

For the argument of the application (i.e., `x_1`), we apply the identity renaming and get the same term again.

More interesting is the function in the application; i.e., `\y_2 -> \x_1 -> x_1`. First, we get to the abstraction for `y_2`, we look it up in `renaming` and, again, find no entry. Hence, we leave that binder as it is and add `2 |-> 2` to `renaming`. 

Next, we traverse `\x_1 -> x_1` with `renaming = {1 |-> 1, 2 |-> 2}`. For the binder `x_1`, we look up `1` in `renaming` and find it exists. Hence, we need a new unique, namely `3`. (We can't just `1 + 1` as we would get a unique that we are already using.)

Now, we change the binder from `x_1 ->` to `x_3 ->` and add `1 |-> 3` to `renaming`. This overwrites the previous mapping for `1`, but that is ok as it reflects the shadowing of the outermost binding of `x`.

Finally, we handle the innermost expression `x_1` with `renaming = {1 |-> 3, 2 |-> 2}`. Looking up `1`, we replace `x_1` by `x_3`, and overall we get

```
\x_1 -> (\y_2 -> \x_3 -> x_3) x_1
``` 

which is exactly what we wanted — all in a single downwards sweep of the AST pushing `renaming` down each branch and adding/updating a mapping with each abstraction.

We may actually prefer to maintain a map `IntMap (Name a)`. Then, you don't need to take apart names and replaces uniques in usage occurrences of variables, but you can just wholesale replace the name by whatever you look up for the unique.